### DISCUSSION TITLE
[torch] Retain a Tensor object past enclosing PointerScope?

### DISCUSSION THREAD

**Discussion by guangster** [bytedeco/javacpp-presets#1160] [Q&A] [answered]
Hi, I would like to persist a Tensor object outside a PoinsterScope. Following [your example here](https://github.com/bytedeco/javacpp/blob/master/src/test/java/org/bytedeco/javacpp/PointerTest.java#L1062-L1071)
gives me access to Tensor but the underlying data seems to get cleaned. The only solution I found is to call `retainReference()` on the original Pointer I passed to `torch.from_blob(Pointer,...)`. This is not ideal as I have to track each Pointer I used to create the Tensor. Neither calling `retainReference()` on Tensor, nor on the output of `Tensor.data_ptr_float()` worked (corrupted data after) even though `Tensor.data_ptr_float()` returned a pointer with the same address as the original Pointer.  Is there a way for us to get the original Pointer from the tensor, or some better ways to retain the tensor?

I attached my example below along with the output I observed. Any help is greatly appreciated!
```java

import org.bytedeco.javacpp.FloatPointer;
import org.bytedeco.javacpp.PointerScope;
import org.bytedeco.pytorch.Tensor;
import org.bytedeco.pytorch.global.torch;

import java.util.Arrays;

public class PointerScopeTest {
    public static void main(String[] args) {
        Tensor tensor;
        FloatPointer pointer;
        float[] buf = new float[3];

        try (PointerScope scope = new PointerScope()) {

            FloatPointer fp = new FloatPointer(new float[]{3, 2, 1});
            tensor = torch.from_blob(fp);
            pointer = new FloatPointer(new float[]{3, 2, 1});
            tensor.retainReference();
            pointer.retainReference();
            // fp.retainReference(); // I would like to not have to call this
            // because I would like to only keep track of tensor.

            
            // work with tensor here...
            // ...


            // sanity check, data still look ok here
            pointer.get(buf);
            System.out.println(Arrays.toString(buf)); //[3.0, 2.0, 1.0]
            tensor.data_ptr_float().get(buf);
            System.out.println(Arrays.toString(buf)); //[3.0, 2.0, 1.0]
        }

        // I would like to be able to access tensor here.
        // It seems the objects are available.
        System.out.println(tensor.isNull()); // false
        System.out.println(pointer.isNull()); // false

        // for Pointer, the data is still ok:
        pointer.get(buf);
        System.out.println(Arrays.toString(buf)); //[3.0, 2.0, 1.0]

        // for Tensor however, the data appears corrupted if I comment out "fp.retainReference();" above
        tensor.data_ptr_float().get(buf);
        System.out.println(Arrays.toString(buf)); //[8.464593E-5, 4.5852E-41, 7.964554E-5]


    }
}
```

**Comment by saudet** [comment, marked as answer]
You could copy the data to a new Tensor instead. You could also transfer the ownership of the Pointer to the new Tensor by providing a deallocator to from_blob().

**Reply by guangster** [reply]
How can we pass a deallocator to from_blob()?Is this the same as [torch.Deleter](http://bytedeco.org/javacpp-presets/pytorch/apidocs/org/bytedeco/pytorch/presets/torch.Deleter.html)

**Reply by guangster** [reply]
I'm not sure how to use this, are you able to provide a brief example how would the syntax look like? The Argument for constructing a torch.Deleter object is just a Pointer, but I'm not sure what to pass?

**Reply by saudet** [reply]
It's a `FunctionPointer`, we need to override the `call()` method:
http://bytedeco.org/javacpp/apidocs/org/bytedeco/javacpp/FunctionPointer.html

But I've also added raw overloads in commit https://github.com/bytedeco/javacpp-presets/commit/20fd80d0eef650a5ca926a302fdfcd7b58b47af6, which should let us pass directly values such as `Pointer.NativeDeallocator.deallocatorAddress` or `Loader.addressof("free")`, if you'd like to give that a try.

**Reply by guangster** [reply]
Thanks! Your new commit worked with `Loader.addressof("free")`. I can now call `retainReference` on both the tensor and the pointer, then just close the tensor to close the data as well. Out of curiosity, how would we pass directly values like `Pointer.NativeDeallocator.deallocatorAddress` since it is a private field? Would we have to use reflection to pull the field out, or is there a better way?

Finally, I attached how I am using your commit below, let me know if I missed anything.
Thanks again!

```java
  public static void main(String[] args) {
        Tensor tensor1, tensor2;
        FloatPointer hidden1, hidden2;
        float[] buf = new float[3];
        float[] data = new float[]{3,2,1};

        try (PointerScope scope = new PointerScope()) {

            hidden1 = new FloatPointer(data);
            hidden2 = new FloatPointer(data);

            tensor1 = torch.for_blob(hidden1).deleter(Loader.addressof("free"))
                    .make_tensor();
            tensor2 = torch.for_blob(hidden2).make_tensor();

            tensor1.retainReference();
            tensor2.retainReference();
            hidden1.retainReference();
            hidden2.retainReference();

        }
        hidden1.get(buf);
        System.out.println("tensor1's data before cleanup:" + Arrays.toString(buf));

        hidden2.get(buf);
        System.out.println("tensor2's data before cleanup:" + Arrays.toString(buf));

        tensor1.close();
        tensor2.close();
        
        hidden1.get(buf);
        System.out.println("tensor1's data after cleanup:" + Arrays.toString(buf));

        hidden2.get(buf);
        System.out.println("tensor2's data after cleanup:" + Arrays.toString(buf));
        
        //tensor1's data before cleanup:[3.0, 2.0, 1.0]
        //tensor2's data before cleanup:[3.0, 2.0, 1.0]
        //tensor1's data after cleanup:[0.0, 0.0, -1.0844938E-19]
        //tensor2's data after cleanup:[3.0, 2.0, 1.0]

    }
```

**Reply by saudet** [reply]
Right, it wasn't intended to be used that way, but I don't see how adding getters could harm anything, so I've done that in commit https://github.com/bytedeco/javacpp/commit/730ae11ecd500fbc483feea493f1ecf719bcd511. With that, it can be accessed this way now:
```java
    class MyFloatPointer extends FloatPointer {
        long deallocatorAddress() {
            return ((NativeDeallocator)deallocator()).deallocatorAddress();
        }
    }
```

If you want to keep using `free()` though, that's only guaranteed to work with memory allocated by `malloc()` and friends, so in that case you should use `Pointer.malloc()` to allocate memory, as done in the samples for Triton Inference Server: https://github.com/bytedeco/javacpp-presets/tree/master/tritonserver/samples

BTW, call `deallocate(false)` instead of `retainReference()`. With the latter, the deallocator may still get called by the GC.

**Reply by guangster** [reply]
You are right, I noticed my Tensor objects were getting garbage collected when I expose them outside Java (i.e. to Python for https://github.com/bytedeco/javacpp/discussions/530), but they worked okay in Java. I ended up setting `org.bytedeco.javacpp.nopointergc` to `true` and that solved it.

I took a look at `deallocate(false)` and logically that makes sense what it would do, but in practice, this didn't quite work for torch's tensor object, as this makes the `data_ptr` object inaccessible (`NullPointerException: This pointer address is NULL`). The Tensor is still fine if I call `deallocate(false)` on `hidden` and `retainReference()` on `tensor` and I can operate with it. I can even call `item_float()` to look at the first element which looks okay, but I cannot get the result out to Java by accessing its `data_ptr_float()`. Only way I found to get the data out is if I call `torch.clone(tensor2).data_ptr_float()`, which shows me that it is working as intended (data is okay outside scope) but not really suitable for my usecase.
Let me know if I have missed something, but if not, it is still not a problem for me because using `nopointergc` works well for me.

**Reply by saudet** [reply]
You need to call deallocate(false) on the original FloatPointer object that has a deallocator. It's not going to work if you call it on the Tensor or on Tensor.data_ptr().

**Reply by saudet** [reply]
It's actually also necessary to call retainReference() since it doesn't remove the deallocator, but only disables it for the GC, so close() can still call it.
